dbos/_core/system_database.py

```diff
@@ -0,0 +1,1734 @@
+        else:
+            async with self.async_engine.begin() as c:
+                results = await c.execute(cmd)
+        if in_recovery:
+            row = results.fetchone()
+            if row is not None:
+                recovery_attempts: int = row[0]
+                if recovery_attempts > max_recovery_attempts:
+                    async with self.async_engine.begin() as c:
+                        await c.execute(
+                            sa.update(SystemSchema.workflow_status)
+                            .where(
+                                SystemSchema.workflow_status.c.workflow_uuid
+                                == status["workflow_uuid"]
+                            )
+                            .where(
+                                SystemSchema.workflow_status.c.status
+                                == WorkflowStatusString.PENDING.value
+                            )
+                            .values(
+                                status=WorkflowStatusString.RETRIES_EXCEEDED.value,
+                            )
+                        )
+                    raise DBOSDeadLetterQueueError(
+                        status["workflow_uuid"], max_recovery_attempts
+                    )
+
+        # Record we have exported status for this single-transaction workflow
+        if status["workflow_uuid"] in self._temp_txn_wf_ids:
+            self._exported_temp_txn_wf_status.add(status["workflow_uuid"])
+
+    def _set_workflow_status(
+        self,
+        conn: sa.Connection,
+        workflow_uuid: str,
+        status: WorkflowStatusString,
+        reset_recovery_attempts: bool,
+    ) -> None:
+        stmt = (
+            sa.update(SystemSchema.workflow_status)
+            .where(SystemSchema.workflow_inputs.c.workflow_uuid == workflow_uuid)
+            .values(
+                status=status,
+            )
+        )
+        conn.execute(stmt)
+
+        if reset_recovery_attempts:
+            stmt = (
+                sa.update(SystemSchema.workflow_status)
+                .where(SystemSchema.workflow_inputs.c.workflow_uuid == workflow_uuid)
+                .values(recovery_attempts=reset_recovery_attempts)
+            )
+            conn.execute(stmt)
+
+    def set_workflow_status_sync(
+        self,
+        workflow_uuid: str,
+        status: WorkflowStatusString,
+        reset_recovery_attempts: bool,
+    ) -> None:
+        with self.sync_engine.begin() as c:
+            self._set_workflow_status(c, workflow_uuid, status, reset_recovery_attempts)
+
+    async def set_workflow_status_async(
+        self,
+        workflow_uuid: str,
+        status: WorkflowStatusString,
+        reset_recovery_attempts: bool,
+    ) -> None:
+        async with self.async_engine.begin() as c:
+            await c.run_sync(
+                self._set_workflow_status,
+                workflow_uuid,
+                status,
+                reset_recovery_attempts,
+            )
```

Comment:
The case above looks like it could have been changed to create the statement in the function, and that running the statement async or sync against the connection.

Reply:
Yes, I suspect most if not all of the `_some_func` / `some_func_sync` / `some_func_async` pattern methods could be implemented that way. I can change these to work that way if we'd rather